tests/Unit/Evolution/Actions/Test_AddMeshVelocityNonconservative.cpp

```diff
@@ -0,0 +1,159 @@
+      db::add_tag_prefix<Tags::dt,
+                         typename Metavariables::system::variables_tag>,
+      domain::Tags::MeshVelocity<Metavariables::volume_dim>>;
+  using phase_dependent_action_list = tmpl::list<
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Initialization,
+          tmpl::list<ActionTesting::InitializeDataBox<simple_tags>>>,
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Testing,
+          tmpl::list<evolution::Actions::AddMeshVelocityNonconservative>>>;
+};
+
+template <typename System>
+struct Metavariables {
+  static constexpr size_t volume_dim = System::volume_dim;
+  using component_list = tmpl::list<component<Metavariables>>;
+  using system = System;
+  enum class Phase { Initialization, Testing, Exit };
+};
+
+template <bool HasMeshVelocity, size_t Dim, typename GradientsTags>
+void test() {
+  using system = System<Dim, GradientsTags>;
+  using metavars = Metavariables<system>;
+  using MockRuntimeSystem = ActionTesting::MockRuntimeSystem<metavars>;
+
+  boost::optional<tnsr::I<DataVector, Dim, Frame::Inertial>> frame_velocity{};
+  if (HasMeshVelocity) {
+    frame_velocity = tnsr::I<DataVector, Dim, Frame::Inertial>{};
+    for (size_t d = 0; d < Dim; ++d) {
+      frame_velocity->get(d) = DataVector{(1. + d) * 5., (1. + d) * 6.};
+    }
+  }
+
+  db::item_type<db::add_tag_prefix<Tags::dt, typename system::variables_tag>>
+      dt_vars(2, 0.);
+  using deriv_tag =
+      db::add_tag_prefix<::Tags::deriv, typename system::variables_tag,
+                         tmpl::size_t<Dim>, Frame::Inertial>;
+  db::item_type<deriv_tag> deriv_vars(2);
+  for (size_t i = 0; i < Dim; ++i) {
+    get<::Tags::deriv<Var1, tmpl::size_t<Dim>, Frame::Inertial>>(deriv_vars)
+        .get(i) = (1. + i) * 3.;
+    for (size_t d = 0; d < Dim; ++d) {
+      get<::Tags::deriv<Var2<Dim>, tmpl::size_t<Dim>, Frame::Inertial>>(
+          deriv_vars)
+          .get(d, i) = (1. + d) * (5. + i);
+    }
+  }
+
+  using simple_tags = typename component<metavars>::simple_tags;
+  MockRuntimeSystem runner{{}};
+  ActionTesting::emplace_component_and_initialize<component<metavars>>(
+      &runner, 0, {deriv_vars, dt_vars, frame_velocity});
+  ActionTesting::set_phase(make_not_null(&runner), metavars ::Phase::Testing);
+  ActionTesting::next_action<component<metavars>>(make_not_null(&runner), 0);
+
+  const auto& box =
+      ActionTesting::get_databox<component<metavars>, simple_tags>(runner, 0);
+
+  for (size_t i = 0; i < Dim; ++i) {
+    DataVector expected{2, 0.};
+    if (HasMeshVelocity) {
+      for (size_t d = 0; d < Dim; ++d) {
+        expected +=
+            frame_velocity->get(d) *
+            get<::Tags::deriv<Var2<Dim>, tmpl::size_t<Dim>, Frame::Inertial>>(
+                deriv_vars)
+                .get(d, i);
+      }
+    }
+    CHECK(db::get<Tags::dt<Var2<Dim>>>(box).get(i) == expected);
```

Comment:
I'm a little surprised that exact equality ends up working out, but if it works on your machine and it works on the CI then perhaps we can expect it to be fine for these simple expressions? I guess it probably only works because the loops compute things in identical orders as in the action.

Reply:
Yea, given how simple the operations are (`Dim` additions) with predictable order the match should generally be exact, but I also don't want to have to deal with it if it breaks so I'll switch `approx` :)